File: implementations/python/tools/perf_annotations.py
#!/usr/bin/env python3
from __future__ import annotations

import statistics
import sys
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
SRC = ROOT / "src"
if str(SRC) not in sys.path:
    sys.path.insert(0, str(SRC))

from aeon.annotations import build_annotation_stream  # noqa: E402
from aeon.core import compile_source  # noqa: E402


SIZES = (1000, 2000, 4000, 8000)
ITERATIONS = 5
RATIO_LIMIT = 2.6


def make_comment_rich(count: int) -> str:
    lines = ["/# top doc #/"]
    for index in range(count):
        lines.append(f"k{index}:number = {index} /? hint-{index} ?/")
    return "\n".join(lines) + "\n"


def median_runtime_seconds(source: str) -> float:
    compiled = compile_source(source)
    if compiled.errors:
        raise RuntimeError(f"fixture failed to compile: {[error.code for error in compiled.errors]}")

    samples: list[float] = []
    for _ in range(ITERATIONS):
        start = time.perf_counter()
        build_annotation_stream(source, compiled.events)
        samples.append(time.perf_counter() - start)
    return statistics.median(samples)


def main() -> int:
    results: list[tuple[int, int, float]] = []
    for size in SIZES:
        source = make_comment_rich(size)
        elapsed = median_runtime_seconds(source)
        results.append((size, len(source.encode("utf-8")), elapsed))

    print("Python annotation perf regression check")
    print("size\tbytes\tmedian_ms\tratio_vs_prev")
    previous: float | None = None
    violations: list[str] = []
    for size, byte_count, elapsed in results:
        ratio_text = "-"
        if previous is not None:
            ratio = elapsed / previous if previous > 0 else float("inf")
            ratio_text = f"{ratio:.2f}"
            if ratio > RATIO_LIMIT:
                violations.append(f"{size} comments grew {ratio:.2f}x over previous size")
        print(f"{size}\t{byte_count}\t{elapsed * 1000:.3f}\t{ratio_text}")
        previous = elapsed

    if violations:
        print("\nFAIL: annotation runtime scaling exceeded threshold", file=sys.stderr)
        for violation in violations:
            print(f"- {violation}", file=sys.stderr)
        return 1

    print(f"\nOK: all growth ratios stayed <= {RATIO_LIMIT:.2f}x")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: implementations/typescript/packages/annotation-stream/scripts/perf-annotations.mjs
import { performance } from 'node:perf_hooks';
import { buildAnnotationStreamFromSource } from '../dist/index.js';
import { compile } from '../../core/dist/index.js';

const SIZES = [1000, 2000, 4000, 8000];
const ITERATIONS = 5;
const RATIO_LIMIT = 2.6;

function makeCommentRich(count) {
  const lines = ['/# top doc #/'];
  for (let index = 0; index < count; index += 1) {
    lines.push(`k${index}:number = ${index} /? hint-${index} ?/`);
  }
  return `${lines.join('\n')}\n`;
}

function median(values) {
  const sorted = [...values].sort((left, right) => left - right);
  const mid = Math.floor(sorted.length / 2);
  return sorted[mid];
}

function medianRuntimeMs(source) {
  const compiled = compile(source, { emitAnnotations: false });
  if ((compiled.errors?.length ?? 0) > 0) {
    throw new Error(`fixture failed to compile: ${compiled.errors.map((error) => error.code).join(',')}`);
  }

  const samples = [];
  for (let iteration = 0; iteration < ITERATIONS; iteration += 1) {
    const start = performance.now();
    buildAnnotationStreamFromSource(source, compiled.events);
    samples.push(performance.now() - start);
  }
  return median(samples);
}

const results = SIZES.map((size) => {
  const source = makeCommentRich(size);
  return {
    size,
    bytes: Buffer.byteLength(source, 'utf8'),
    medianMs: medianRuntimeMs(source),
  };
});

console.log('TypeScript annotation perf regression check');
console.log('size\tbytes\tmedian_ms\tratio_vs_prev');

let previous = null;
const violations = [];
for (const result of results) {
  let ratioText = '-';
  if (previous !== null) {
    const ratio = previous > 0 ? result.medianMs / previous : Number.POSITIVE_INFINITY;
    ratioText = ratio.toFixed(2);
    if (ratio > RATIO_LIMIT) {
      violations.push(`${result.size} comments grew ${ratio.toFixed(2)}x over previous size`);
    }
  }
  console.log(`${result.size}\t${result.bytes}\t${result.medianMs.toFixed(3)}\t${ratioText}`);
  previous = result.medianMs;
}

if (violations.length > 0) {
  console.error(`\nFAIL: annotation runtime scaling exceeded threshold (${RATIO_LIMIT.toFixed(2)}x)`);
  for (const violation of violations) {
    console.error(`- ${violation}`);
  }
  process.exit(1);
}

console.log(`\nOK: all growth ratios stayed <= ${RATIO_LIMIT.toFixed(2)}x`);
